tests/unit/s2n_client_hello_request_test.c

```diff
@@ -108,23 +114,72 @@ int main(int argc, char **argv)
         EXPECT_SUCCESS(s2n_negotiate_test_server_and_client(server_conn, client_conn));
     }
 
-    /* Test: Hello request received after the handshake */
-    {
+    /* Test: Hello request not accepted for TLS1.3 */
+    if (s2n_is_tls13_fully_supported()) {
         DEFER_CLEANUP(struct s2n_connection *client_conn = s2n_connection_new(S2N_CLIENT),
                 s2n_connection_ptr_free);
         EXPECT_NOT_NULL(client_conn);
+        EXPECT_SUCCESS(s2n_connection_set_cipher_preferences(client_conn, "default_tls13"));
         EXPECT_SUCCESS(s2n_connection_set_config(client_conn, config));
 
         DEFER_CLEANUP(struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER),
                 s2n_connection_ptr_free);
         EXPECT_NOT_NULL(server_conn);
+        EXPECT_SUCCESS(s2n_connection_set_cipher_preferences(server_conn, "default_tls13"));
         EXPECT_SUCCESS(s2n_connection_set_config(server_conn, config));
 
         DEFER_CLEANUP(struct s2n_test_io_pair io_pair = { 0 }, s2n_io_pair_close);
         EXPECT_SUCCESS(s2n_io_pair_init_non_blocking(&io_pair));
         EXPECT_SUCCESS(s2n_connection_set_io_pair(client_conn, &io_pair));
         EXPECT_SUCCESS(s2n_connection_set_io_pair(server_conn, &io_pair));
 
+        /* Start the handshake */
+        EXPECT_OK(s2n_negotiate_test_server_and_client_until_message(server_conn, client_conn,
+                SERVER_HELLO));
+
+        /* Send a hello request.
+         * We should be able to receive the hello request before the version is negotiated. */
+        EXPECT_OK(s2n_send_client_hello_request(server_conn));
+
+        /* Continue the handshake */
+        EXPECT_OK(s2n_negotiate_test_server_and_client_until_message(server_conn, client_conn,
+                ENCRYPTED_EXTENSIONS));
+
+        /* Send a hello request.
+         * We should NOT be able to receive the hello request after TLS1.3 is negotiated */
+        EXPECT_EQUAL(server_conn->actual_protocol_version, S2N_TLS13);
+        EXPECT_EQUAL(client_conn->actual_protocol_version, S2N_TLS13);
+        EXPECT_OK(s2n_send_client_hello_request(server_conn));
+
+        /* Handshake should fail because the HelloRequest is unexpected in TLS1.3 */
+        EXPECT_FAILURE_WITH_ERRNO(s2n_negotiate_test_server_and_client(server_conn, client_conn),
+                S2N_ERR_BAD_MESSAGE);
+    }
+
+    /* Test: Hello request received after the handshake */
+    {
```

Comment:
is this testing that if behavior is `S2N_ALERT_IGNORE_WARNINGS`, then `no_renegotiation` alert is not fatal. could you state that in the description here.

Reply:
We're not actually trying to test the S2N_ALERT_IGNORE_WARNINGS behavior; that's tested elsewhere. We're more interested in whether the connection can continue after the HelloRequest message. We just need S2N_ALERT_IGNORE_WARNINGS because s2n-tls is very strict about alerts by default.

I reworked all the test names-- see if that's any clearer.

Reply:
I guess to be clearer, we're not super interested in how the warning is handled here because the warning is only received if a server requests renegotiation, which the s2n-tls server will never actually do. Note that I had to manually inject the "server"'s HelloRequest message :) We're primarily testing how the HelloRequest message is received, meaning we're primarily interested in the client.